src/datasources/asset/AssetDataSource.ts

```diff
@@ -0,0 +1,325 @@
+    readonly instanceSettings: DataSourceInstanceSettings,
+    readonly backendSrv: BackendSrv = getBackendSrv(),
+    readonly templateSrv: TemplateSrv = getTemplateSrv()
+  ) {
+    super(instanceSettings, backendSrv, templateSrv);
+  }
+
+  baseUrl = this.instanceSettings.url + '/niapm/v1';
+  sysmgmtUrl = this.instanceSettings.url + '/nisysmgmt/v1'
+
+  defaultQuery = {
+    assetQueryType: AssetQueryType.METADATA,
+    workspace: '',
+    entityType: EntityType.ASSET,
+    assetIdentifier: '',
+    isNIAsset: IsNIAsset.BOTH,
+    minionId: '',
+    isPeak: IsPeak.PEAK,
+    timeFrequency: TimeFrequency.DAILY,
+    utilizationCategory: UtilizationCategory.ALL,
+    peakDays: [Weekday.Monday, Weekday.Tuesday, Weekday.Wednesday, Weekday.Thursday, Weekday.Friday],
+    policyOption: PolicyOption.DEFAULT,
+    peakStart: undefined,
+    nonPeakStart: undefined
+  };
+
+  async runQuery(query: AssetQuery, options: DataQueryRequest): Promise<DataFrameDTO> {
+    const result: DataFrameDTO = { refId: query.refId, fields: [] };
+
+    if (query.assetQueryType === AssetQueryType.METADATA) {
+      let workspace_id = this.templateSrv.replace(query.workspace);
+      let minion_id = this.templateSrv.replace(query.minionId);
+      const conditions = [];
+      if (workspace_id) {
+        conditions.push(`workspace = "${workspace_id}"`);
+      }
+      if (minion_id) {
+        conditions.push(`${AssetFilterProperties.LocationMinionId} = "${minion_id}"`);
+      }
+      const assetFilter = conditions.join(' and ');
+      let assetResponse: AssetModel[] = await this.queryAssets(assetFilter);
+      result.fields = [
+        { name: 'id', values: assetResponse.map(a => a.id) },
+        { name: 'name', values: assetResponse.map(a => a.name) },
+        { name: 'model name', values: assetResponse.map(a => a.modelName) },
+        { name: 'serial number', values: assetResponse.map(a => a.serialNumber) },
+        { name: 'bus type', values: assetResponse.map(a => a.busType) },
+        { name: 'asset type', values: assetResponse.map(a => a.assetType) },
+        { name: 'is NI asset', values: assetResponse.map(a => a.isNIAsset) },
+        { name: 'calibration status', values: assetResponse.map(a => a.calibrationStatus) },
+        { name: 'is system controller', values: assetResponse.map(a => a.isSystemController) },
+        { name: 'workspace', values: assetResponse.map(a => a.workspace) },
+        { name: 'last updated timestamp', values: assetResponse.map(a => a.lastUpdatedTimestamp) },
+        { name: 'minionId', values: assetResponse.map(a => a.location.minionId) },
+        { name: 'parent name', values: assetResponse.map(a => a.location.parent) },
+        { name: 'system name', values: assetResponse.map(a => a.location.systemName) },
+        {
+          name: 'calibration due date',
+          values: assetResponse.map(a => a.externalCalibration?.nextRecommendedDate)
+        },
+      ];
+      return result;
+    } else {
```

Comment:
Can we break this up into at least 2 methods, one for the metadata and one for the utilization? This function is really long.